feat(list): supplement some list cmds, and coresponding tests

MciG-ggg opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Added four list operations: LPUSHX/RPUSHX, LINSERT (BEFORE/AFTER), and RPOPLPUSH; exposed a Before/After selector.

* **Bug Fixes**
  * Improved per-key locking and concurrency safety, consistent empty-result behavior, safe pop limits, and cross-key validation for pop-and-push.

* **Tests**
  * Expanded coverage for new list commands, concurrency, atomicity, and many edge cases.

* **Chores**
  * Updated repository ignore patterns.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

marsevilspirit reviewed (approved):
LGTM

Commit:
fix(cmd): add some boundary check for list cmd

Commit:
fix(storage): fix LPUSHX/RPUSHX behavior on empty lists

- Refactor push_core function to unify lpush/rpush/lpushx/rpushx operations
- Fix empty list metadata preservation to allow pushx operations on existing empty lists
- Fix empty values array edge case to return current length for existing lists instead of 0
- Comply with Redis protocol: lpushx/rpushx only operate when key exists, otherwise return 0

Impact:
- src/storage/src/redis_lists.rs: push_core function refactoring
- lpop/rpop methods: preserve empty list metadata for subsequent pushx operations
- All list operations now handled through unified push_core function

Commit:
feat(cmd): add LPUSHX and RPUSHX command support

- Add LPushxCmd and RPushxCmd command implementations
- Add lpushx/rpushx methods to Storage trait
- Implement storage layer calls in storage_impl.rs
- Support Redis protocol conditional push operations: execute only when list exists

New features:
- src/cmd/src/list.rs: LPushxCmd and RPushxCmd command structures
- src/storage/src/storage_impl.rs: trait methods and implementations
- Complete error handling and response formatting

Commit:
test(storage): enhance Redis list operations test cases

- Add dedicated LPUSHX and RPUSHX tests: empty list operations, edge cases
- Add mixed operations tests: verify lpushx/rpushx compatibility with other list ops
- Improve formatting and comments for existing test cases
- Cover all edge cases: empty values array, empty lists, non-existent keys

New tests:
- test_lpushx_on_empty_list: verify lpushx operation on empty lists
- test_rpushx_on_empty_list: verify rpushx operation on empty lists
- test_lpushx_rpushx_edge_cases: edge cases and mixed operations test
- test_lpushx_rpushx_mixed_operations: complex mixed operation scenarios

All 20 test cases now pass, ensuring list operation correctness and stability.

Commit:
chore: add Serena AI config directory to .gitignore

- Ignore .serena/ directory to avoid committing AI assistant configuration files
- Keep repository clean with only project source code and essential configs

Commit:
feat: improve Redis list operations and add boundary checks

Commit:
fix: resolve compilation errors in Redis list commands

- Fix BeforeOrAfter enum import path by properly exporting from storage module
- Fix string literal type errors by adding byte string prefixes (b"BEFORE"/b"AFTER")
- Fix RespData::BulkString type mismatch by wrapping Vec<u8> in Some(Bytes::from())
- Replace RespData::NullBulkString with RespData::BulkString(None) for consistency
- Remove unused fields from LInsertCmd struct to eliminate dead code warnings
- Update test imports and references to use correct BeforeOrAfter path
- Fix error handling in storage_impl.rs to use proper snafu error pattern

All compilation errors resolved and redis_list_test suite passes successfully.

Commit:
fix: resolve compilation errors in Redis list commands

- Fix BeforeOrAfter enum import path by properly exporting from storage module
- Fix string literal type errors by adding byte string prefixes (b"BEFORE"/b"AFTER")
- Fix RespData::BulkString type mismatch by wrapping Vec<u8> in Some(Bytes::from())
- Replace RespData::NullBulkString with RespData::BulkString(None) for consistency
- Remove unused fields from LInsertCmd struct to eliminate dead code warnings
- Update test imports and references to use correct BeforeOrAfter path
- Fix error handling in storage_impl.rs to use proper snafu error pattern

All compilation errors resolved and redis_list_test suite passes successfully.

Commit:
refactor(cmd): improve linsert and rpoplpush method signatures for clarity and consistency

Commit:
refactor(cmd): improve linsert and rpoplpush method signatures for clarity and consistency

Commit:
fix(cmd): change LPushx and RPushx 's arguments from 3 to at least 3

Commit:
fix(cmd): replace clone() with to_vec() for slice to Vec conversion

- Update LPushxCmd and RPushxCmd to use to_vec() instead of clone()
- Improves clarity of conversion from slice to Vec
- Complements previous commit 730fc8a that added multi-value support

guozhihao-224 commented:
It appears that no key locking operation was performed, which could lead to data corruption in the metadata.

MciG-ggg reviewed (commented):
> It appears that no key locking operation was performed, which could lead to data corruption in the metadata. 

So should i add rwlock or sth else to the concurrent critical section?

Commit:
fix(cmd): add ScopeRecordLock to list cmd

Commit:
Merge branch 'main' into main

Commit:
fix(cmd):fix the deadlock problem of rpoplpush

guozhihao-224 commented on src/storage/src/storage_impl.rs:
```diff
@@ -422,6 +440,31 @@ impl Storage {
+    pub fn rpoplpush(&self, source_key: &[u8], destination_key: &[u8]) -> Result<Option<Vec<u8>>> {
+        let source_slot_id = key_to_slot_id(source_key);
+        let dest_slot_id = key_to_slot_id(destination_key);
+        if source_slot_id != dest_slot_id {
```
- Cross-slot errors should not be exposed to the client; the implementation in CPP can be referenced： https://github.com/arana-db/kiwi-cpp/blob/0fa4f1b95ff2f89873d06eb86de40532dd91d146/src/storage/src/storage.cc#L946 
- However, I believe the CPP implementation also has problems. The semantics of the RPOPLUSH command are atomic; the involved keys should be locked when the command is issued to ensure atomicity.

MciG-ggg replied:
1. Ok, i get it, i will fix it
2. I have read the implementation of CPP version, i see that it split RPoplpush to RPop and LPush, and it does not lock any of these two commands. But in my implementation, in my opinion, i lock both commands, you can check it at https://github.com/MciG-ggg/kiwi/blob/main/src/storage/src/redis_lists.rs#L1021

guozhihao-224 committed:
feat(storage): add data and meta compaction filters (#183)

* feat(storage): add data and meta compaction filters

guozhihao-224 committed:
feat(storage): enhance data type handling and add expiration check (#190)

* feat(storage): enhance data type handling and add expiration check

- Introduced `min_meta_raw_len` method in `DataType` enum to calculate minimum metadata length for different data types.
- Added constants for list metadata value length in `list_meta_value_format.rs`.
- Implemented `is_stale` method in `Redis` struct to check if encoded values are expired, improving data integrity checks.
- Updated `BASE_META_VALUE_LENGTH` to reflect changes in metadata structure.

* refactor(storage): update min_meta_raw_len to return Result

- Modified `min_meta_raw_len` method in `DataType` enum to return a `Result<usize>` instead of `usize`, enhancing error handling for invalid data types.
- Updated usage in `Redis` struct to handle the new Result type, ensuring proper error propagation for invalid value lengths and data types.

* test: add test for is_stale

---------

Co-authored-by: Xin.Zh <alexstocks@foxmail.com>

Commit:
Merge branch 'main' into main

Commit:
fix(cmd): fix some errors

Commit:
Merge branch 'arana-db:main' into main

Commit:
Merge branch 'arana-db:main' into main

Commit:
Merge branch 'arana-db:main' into main

Commit:
Merge branch 'main' into main

Commit:
fix: add BeforeOrAfter

Commit:
fix: make fmt
